presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -327,21 +374,26 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
             builder.add((new IterativeOptimizer(
                     stats,
                     statsCalculator,
+                    estimatedExchangesCostCalculator,
                     ImmutableList.of(new com.facebook.presto.sql.planner.optimizations.DetermineJoinDistributionType()), // Must run before AddExchanges
                     ImmutableSet.of(new DetermineJoinDistributionType())))); // Must run before AddExchanges
             builder.add(new DetermineSemiJoinDistributionType()); // Must run before AddExchanges
             builder.add(
                     new IterativeOptimizer(
                             stats,
                             statsCalculator,
+                            estimatedExchangesCostCalculator,
                             ImmutableSet.of(new PushTableWriteThroughUnion()))); // Must run before AddExchanges
             builder.add(new AddExchanges(metadata, sqlParser));
         }
+        //noinspection UnusedAssignment
+        estimatedExchangesCostCalculator = null; // Prevent accidental use after AddExchanges
```

Comment:
maybe you could extract two methods, one with one cost calculator provided and other with second calculator?

Reply:
no!